Refactor repository scaffolding

tomvothecoder opened the pull request:
## Overview
This PR updates the scaffolding of the repository to follow Python conventions. It also removes the `esgfpub` directory since it is no longer being used.

## Summary of changes
- Move `/warehouse` directory one level down to the root of the repo.
  - This directory was previously `/warehouse/warehouse`.
  - Python module imports were safely updated using PyCharm.
- Rename `/warehouse` directory  to `/datasm`.
  - Python module imports were safely updated using PyCharm.
  - Update some `warehouse` string refs to `datasm`, except paths, command arguments, and classes.
- Remove `/esgfpub` directory and related refs.
  - Closes #104 

DevOps
- Move `/datasm/setup.py` to the root of repo and update contents.
  - Now we run `pip install . ` from the root of the repo, instead of `cd warehouse` first.
- Rename `dev.yml` to `proc.yml`. 
  - Update references for "development" to "processing"
  - It is more accurate to call this environment the "processing" env because the "publishing" env is also used for development.
- Remove `prod.yml` since it is not working yet. 
   - Need to resolve #110 and #99 first

Documentation
- Update `README.md` and `docs/getting_started.md`
- Move `NewCMIPVariableSetup.md` and `warehouse_processing.md` to `/docs`
- Update `.gitignore` with standard list from https://github.com/github/gitignore/blob/main/Python.gitignore

## What to review in this PR
I already performed a self-review of the items below, but more eyes are always. Most of the changes in the diff are related to deleting the `/esgfpub` directory, so you can ignore those

- [x] Make sure documentation updates are free of error (mostly path updates)
- [x] Local `datasm` build still builds and installs in the processing and publishing envs
- [x] Local `datasm` build runs the same as before using `datasm` command, instead of `warehouse`
  - Example: `datasm --help`, `datasm validate --help`

## How to test this PR

Processing operations -- `extract`, `validate`, `postprocess`
1. Install and activate `datasm_proc` env
   - `conda env install -f conda-env/proc.yml`
   - `conda activate datasm_proc`
2. Install local build into `datasm_proc` env
   - Make sure you are in the root of `esgfpub`
   - `pip install .`
3. Run `datasm` processing operations to make sure things are good to go
   - For example, confirm there are no `ModuleNotFound` errors

Publishing operations -- `process`
1. Install and activate `datasm_pub` env
   - `conda env install -f conda-env/pub.yml`
   - `conda activate datasm_pub`
2. Install local build into `datasm_pub` env
   - Make sure you are in the root of `esgfpub`
   - `pip install .`
3. Run `datasm` publishing operation to make sure things are good to go
   - For example, confirm there are no `ModuleNotFound` errors

## Next Actions
- [x] Rename the repository to `datasm`

tomvothecoder commented:
I just realized there are merge conflicts with the latest commits, so this actually isn't ready for review yet.

tomvothecoder commented on e3sm_warehouse/e3sm_warehouse.py:
```diff
@@ -654,7 +655,7 @@ def add_args(
         )
         p.add_argument(
             "-w",
-            "--warehouse-path",
```
I might need to revert these specific argument name changes since they probably don't make sense

TonyB9000 replied:
This is another example of term-overloading. We still refer to the user_pub filesystem location, to which archives datasets are extracted, as the "warehouse" (fully, /p/user_pub/e3sm/warehouse), and the "-w" of the command line allows a user to override that location.

TonyB9000 replied:
Note that there is a corresponding argument "-p", indicating the "publication" location of datasets that are published.  Often referred to as the "pub_dirs", that default is "/p/user_pub/work" an historical choice that I think is terrible, but necessary legacy access.

tomvothecoder replied:
Got it, I will revert these types of changes to `warehouse`. `e3sm_warehouse` will refer strictly to the Python package, not any paths.

tomvothecoder commented on e3sm_warehouse/resources/warehouse_config.yaml:
```diff
@@ -1,4 +1,4 @@
-DEFAULT_WAREHOUSE_PATH: /p/user_pub/e3sm/warehouse/
```
Should we rename this directory to align with the repo and package?

TonyB9000 replied:
No, I'd want to avoid that.  It already has e3sm/warehouse in the path, and e3sm/e3sm_warehouse would be redundant.  Also, MANY scripts and config files rely upon it - that would be a lot of editing and testing.

tomvothecoder replied:
Got it, that's what I was thinking too. I will revert my "find and replace" string changes that affected these paths.

tomvothecoder commented on e3sm_warehouse/e3sm_warehouse.py:
```diff
@@ -16,13 +17,13 @@
 from termcolor import colored, cprint
 
 
-from warehouse.workflows import Workflow
-from warehouse.dataset import Dataset, DatasetStatus
-from warehouse.slurm import Slurm
-from warehouse.listener import Listener
-import warehouse.resources as resources
-import warehouse.util as util
-from warehouse.util import setup_logging, log_message
```
Example module import updates

TonyB9000 replied:
This is another example of term-overloading.  We still refer to the user_pub filesystem location, to which archives datasets are extracted, as the "warehouse" (fully, /p/user_pub/e3sm/warehouse), and the "-w" of the command line allows a user to override that location.

TonyB9000 replied:
This is another matter. Granted, we have no other "warehouse" project, but changing to "e3sm_warehouse" is a possibility.  Major hassle, but if it makes something cleaner, it is do-able.

tomvothecoder commented on conda-env/ci.yml:
```diff
@@ -1,6 +1,6 @@
 # Conda continuous integration/continuous deployment (CI/CD) for running the repository's GitHub Actions CI/CD workflows.
 # End-users are not expected to use this environment.
-name: warehouse_ci
```
Added `e3sm_` suffix to conda env yml file names and prefixs

TonyB9000 replied:
I'm not sure of the impact here.  It would not hurt, I suspect.

tomvothecoder replied:
This comment was just to make reviewers aware

tomvothecoder reviewed (commented):
A few comments

chengzhuzhang reviewed (approved):
Thanks for the PR. 
At some point, if Tony can have a test to run warehouse with these changes, I'd like to try as well

TonyB9000 commented:
We do have "datasets" we can "warehouse_extract", "warehouse_validate", "warehouse_generate", etc, that are small and could be good for high-level testing.  But someone will need to walk me through the steps I need to take to make this transition.  This includes any potential re-cloning of repos, any re-generation of conda environments, and then I will need to know what paths have changed (e.g. changing "warehouse" to "e3sm_warehouse") so I can search for them.  A lot of "quickie-hardcoded" stuff in scripts will need to be turned into ENV variables, so defined in one location.

tomvothecoder commented on datasm/dataset.py:
```diff
@@ -270,7 +270,7 @@ def latest_warehouse_dir(self):
         # import ipdb; ipdb.set_trace()
 
-        # we assume that the warehouse directory contains only directories named "v0.#" or "v#"
+        # we assume that the wraehouse directory contains only directories named "v0.#" or "v#"
```
```suggestion
        # we assume that the warehouse directory contains only directories named "v0.#" or "v#"
```

tomvothecoder commented on datasm/warehouse.py:
```diff
@@ -645,7 +646,7 @@ def collect_e3sm_datasets(self, **kwargs):
         parser,
     ):
-        p = parser.add_parser(name=NAME, help="Automated warehouse processing")
+        p = parser.add_parser(name=NAME, help="Automated processing")
```
```suggestion
        p = parser.add_parser(name=NAME, help="Automated warehouse processing")
```

tomvothecoder commented on docs/getting_started.md:
```diff
@@ -109,35 +109,35 @@ Occasionally, the dependencies in the conda env yml files are updated to include
   -s, --serial          Run everything in serial
-  -w WAREHOUSE_PATH, --warehouse-path WAREHOUSE_PATH
-                        The root path for pre-publication dataset staging, default=/p/user_pub/e3sm/warehouse/
+  -w WAREHOUSE_PATH, --datasm-path WAREHOUSE_PATH
```
```suggestion
  -w WAREHOUSE_PATH, --warehouse-path WAREHOUSE_PATH
```

tomvothecoder commented on docs/getting_started.md:
```diff
@@ -109,35 +109,35 @@ Occasionally, the dependencies in the conda env yml files are updated to include
-  -w WAREHOUSE_PATH, --warehouse-path WAREHOUSE_PATH
-                        The root path for pre-publication dataset staging, default=/p/user_pub/e3sm/warehouse/
+  -w WAREHOUSE_PATH, --datasm-path WAREHOUSE_PATH
+                        The root path for pre-publication dataset staging, default=/p/user_pub/e3sm/datasm/
```
```suggestion
                        The root path for pre-publication dataset staging, default=/p/user_pub/e3sm/warehouse/
```

tomvothecoder commented on setup.py:
```diff
@@ -46,23 +44,20 @@ def run(self):
     },
     test_suite="tests",
-    url="https://github.com/XCDAT/xcdat",
+    url="https://github.com/E3SM-Project/esgfpub",
```
```suggestion
    # TODO: Update this URL once the repo is renamed.
    url="https://github.com/E3SM-Project/esgfpub",
```

TonyB9000 reviewed (approved):
This looks good to me.  Thanks Tom.

tomvothecoder commented:
> This looks good to me. Thanks Tom.

Thanks Tony. I pushed some new changes just now, which include renaming the `dev.yml` to `proc.yml`. Please review the latest PR description.

If you have time, it would be nice for you to test the new `datasm` command which supersedes `warehouse`. All of the subcommands should be the same. I provided instructions in the PR's description.

I would also like to merge this PR when it is least disruptive to your workflow, so that is dependent on your priority list.

TonyB9000 commented:
Hi Tom,  Did you want "proc" or "prod"?

I have a SLEW of updates to commit and merge to master.  How should I coordinate that with your repo-refactoring?

(I expect I will wipe/rename the old "esgfpub", and git-clone everything fresh.  THEN worry about external scripts vis-a-vis naming.

tomvothecoder commented:
> Hi Tom, Did you want "proc" or "prod"?
>

@TonyB9000, I removed the `prod.yml` env for now to avoid confusion because a versioned release of this package doesn't exist yet. I also renamed `dev.yml` to `proc.yml`, which is the environment intended for "processing" operations (`extract, `validate,` postprocess`). 

I would like for you to test the local build (`pip install .`) using both the `proc.yml` and `pub.yml` and their related operations/commands.

> > I have a SLEW of updates to commit and merge to master. How should I coordinate that with your repo-refactoring?
> 
> (I expect I will wipe/rename the old "esgfpub", and git-clone everything fresh. THEN worry about external scripts vis-a-vis naming.)
> 

I think we should prioritize merging your changes to `master` first, then rebase this branch on top of `master`. I will update the affected files so that the imports reflect the new `datasm` name.

TonyB9000 commented:
OK, I will do a push and PR, so we can do this cut-over rather than drag it out.  My changes should incur no conflicts (none serious) and most recent are simply enhanced logging.

tomvothecoder commented:
Hey @TonyB9000, this PR is ready for review. Please read through the description for the changelog, what to review, and how to test.

TonyB9000 commented:
Whole Lotta regression testing for name-changes in my "ancillary" codes and scripts - but to be expected...

TonyB9000 commented:
(I'm especially interested in the "How to Test" part.)

tomvothecoder commented:
> How do I find the "changelog, what to review, and how to test"? (and I've much other stuff to test as well.)

The PR's description

TonyB9000 commented:
Excellent!  (You mean I have to read?  ;) ).  OK, here I go.

TonyB9000 commented:
The testing explains "Install and activate datasm_proc env".  But my (esgfpub) conda-env still has "dev.yml", etc.

Should I "git clone https://github.com/E3SM-Project/datasm" first?  Where is conda-env bundled?

tomvothecoder commented:
> The testing explains "Install and activate datasm_proc env". But my (esgfpub) conda-env still has "dev.yml", etc.
and 
> Where is conda-env bundled?

You need to checkout the branch `devops/104-deprecate-esgfpub` to access the new conda env yml files.

> Should I "git clone https://github.com/E3SM-Project/datasm" first? 

The repo is going to be renamed to `datasm` after this PR is merged.

TonyB9000 commented:
OK, thanks.  Should "conda env install -f conda-env/proc.yml" be "conda env create"?

TonyB9000 commented:
Testing: So far so good!  At least, datasm -h, datasm validate -h, and datasm postprocess -h all produce the expected help-text.  Tomorrow I'll conduct some heavy testing.  I recall the "module load" errors were not triggered by running "help", but when loading the datasm subordinate modules with datasets.

tomvothecoder commented:
Awesome, thanks Tony!

tomvothecoder commented:
Hey @TonyB9000, how is testing going for this PR?

TonyB9000 commented:
Hey @tomvothecoder , all is fine with postprocessing (and I suspect anything invoked under "datasm <subprocess>").  I cannot test publication until I am ready to publish, and I have not yet tested many ancillary applications (consolidated status reporting) for naming or path issues - but they should be trivial to find and fix when the time comes.  I have a HUGE backlog of datasets to generate (and then, to publish) so my focus lies there,  By next Monday, I believe ALL of the "Amon" sets I am generating will have completed, and I will test "datasm publish" (and conda env "dsm_pub") at that time.  I can pause on Monday(ish) if you need to interrupt my job sequence.  (I *could* search all of my scripts for the term "warehouse" and see if any references to that module name remain).

TonyB9000 commented:
I have been "testing" (operating) dsm_proc and dsm_pub in the course of getting publication work done.  There is no way to thoroughly test, as there are too many one-off cases of things that need doing.  But I am satisfied that the name-changes and related restructuring have not broken anything serious.  In time, glitches will be uncovered and addressed.

Right now (in the middle of an ongoing set of post-process jobs) my only worry is that I should not "pip install" datasm while the plane in in flight.  In particular, it is calling e3sm_to_cmip routinely.  The current batch of jobs will finish before the Monday.

Perhaps I should (then) hold off starting new jobs until we can eliminate "devops/104-deprecate-esgfpub"?

tomvothecoder commented:
Thanks @TonyB9000. Yes, let's hold off on further jobs until this PR is merged.

We should wait until the current batch of jobs are done Monday before we merge this PR.

tomvothecoder commented:
@TonyB9000 I will go ahead and merge this PR. 

We will address any issues that pop up as we use `datasm`, but hopefully there are none since most changes are directory structure and name changes.

tomvothecoder commented:
@TonyB9000 and @chengzhuzhang the repository has been renamed to "datasm". 

Your current local cloned repos with the previous "esgfpub" directory name will still work, but you will receive a git warning to update the name of the directory to "datasm".

Per https://stackoverflow.com/a/60082006:

> Renaming the Local Repo Name
>
> As others have mentioned, the local "name" of your repo is typically considered to be the root folder/directory name, and you can change that, move, or copy the folder to any location and it will not affect the repo at all.
> 
> Git is designed to only worry about files inside the root folder.
